http/src/v0/pin.rs

```diff
@@ -116,26 +107,224 @@ impl<'a> TryFrom<&'a str> for AddRequest {
             }
         }
 
-        if args.is_empty() {
-            return Err(MissingArg);
-        }
+        // special case compared to others: it's ok not to have any args, or any filter
 
-        Ok(AddRequest {
-            args,
-            recursive: recursive.unwrap_or(false),
-            progress: progress.unwrap_or(false),
+        Ok(ListRequest {
+            arg: args,
+            filter: filter.unwrap_or(PinFilter::All),
+            quiet: quiet.unwrap_or(false),
+            // this default was mentioned in the pin/ls api
+            stream: quiet.unwrap_or(true),
+            timeout,
         })
     }
 }
 
-/// Filter to perform custom `warp::query::<AddRequest>`. This needs to be copypasted around as
-/// HRTB is not quite usable yet.
-fn add_request() -> impl Filter<Extract = (AddRequest,), Error = Rejection> + Clone {
+#[derive(Debug, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum PinFilter {
+    Direct,
+    Indirect,
+    Recursive,
+    All,
+}
+
+impl Default for PinFilter {
+    fn default() -> Self {
+        PinFilter::All
+    }
+}
+
+impl PinFilter {
+    fn to_mode(&self) -> Option<PinMode> {
+        use PinFilter::*;
+        match self {
+            Direct => Some(PinMode::Direct),
+            Indirect => Some(PinMode::Indirect),
+            Recursive => Some(PinMode::Recursive),
+            All => None,
+        }
```

Comment:
nit: seems to me that `PinFilter` and `PinMode` are nearly the same; it's not a big deal, but perhaps a single object like
```
enum PinMode {
    Direct,
    Indirect,
    Recursive,
    Off / Disabled / etc.
}
```
would work for both use cases?

Reply:
Dont forget PinKind. Having these gazillion types is me trying to isolate the http boundary stuff from the ipfs stuff. I think extending PinMode and providing the FromStr would not make much sense... Well, the (last) `All` variant would not be a problem anymore. I used to have PinMode as a parameter for `PinStore::insert_pin` but of course that was replaced with PinKind.

Reply:
Not doing anything for this right now, but keeping an eye out if it's possible to get rid of one, or at least the `pub` modifier while hiding the current impl into mem.rs and working on the file based.